File: ios/MRT/Features/Sessions/SessionViewModel.swift
import Foundation
import SwiftUI

@MainActor
final class SessionViewModel: ObservableObject {
    @Published var sessions: [SessionModel]
    @Published var activeSessionID: String?

    private let connectionManager: ConnectionManaging?

    init(connectionManager: ConnectionManaging? = nil, sessions: [SessionModel]? = nil) {
        self.connectionManager = connectionManager
        let initialSessions = sessions ?? (connectionManager == nil ? Self.defaultSessions : connectionManager?.sessions ?? [])
        self.sessions = initialSessions
        self.activeSessionID = initialSessions.first?.id
        connectionManager?.onSessionsChange = { [weak self] authoritativeSessions in
            Task { @MainActor in
                self?.applyAuthoritativeSessions(authoritativeSessions)
            }
        }
    }

    @discardableResult
    func selectSession(id: String) -> Bool {
        guard sessions.contains(where: { $0.id == id }) else {
            return false
        }

        activeSessionID = id

        guard let connectionManager else {
            return true
        }

        Task {
            try? await connectionManager.switchSession(to: id)
        }

        return true
    }

    func canCreateSession(connectionState: ConnectionState? = nil) -> Bool {
        guard connectionManager != nil else { return true }
        return (connectionState ?? connectionManager?.state) == .connected
    }

    func createSession(named name: String) {
        let trimmedName = name.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmedName.isEmpty else { return }

        if let connectionManager {
            guard canCreateSession() else { return }
            Task {
                try? await connectionManager.createSession(name: trimmedName, workingDirectory: "")
            }
            return
        }

        let session = SessionModel(
            id: UUID().uuidString,
            name: trimmedName,
            status: .idle,
            createdAtMs: Self.nowMilliseconds(),
            lastActiveMs: Self.nowMilliseconds(),
            workingDirectory: "/tmp/\(trimmedName.replacingOccurrences(of: " ", with: "-").lowercased())"
        )
        sessions.insert(session, at: 0)
        activeSessionID = session.id
    }

    func closeSession(id: String) {
        if let connectionManager {
            guard sessions.contains(where: { $0.id == id }) else { return }
            Task {
                try? await connectionManager.closeSession(id: id)
            }
            return
        }

        sessions.removeAll { $0.id == id }
        if activeSessionID == id {
            activeSessionID = sessions.first?.id
        }
    }

    func cancelTask(id: String) {
        if let connectionManager {
            guard sessions.contains(where: { $0.id == id }) else { return }
            Task {
                try? await connectionManager.cancelTask(sessionID: id)
            }
            return
        }

        let updatedAt = Self.nowMilliseconds()
        sessions = sessions.map { session in
            guard session.id == id else { return session }
            return SessionModel(
                id: session.id,
                name: session.name,
                status: .cancelled,
                createdAtMs: session.createdAtMs,
                lastActiveMs: updatedAt,
                workingDirectory: session.workingDirectory
            )
        }
    }

    private static var defaultSessions: [SessionModel] {
        [
            SessionModel(
                id: "session-1",
                name: "Main Session",
                status: .running,
                createdAtMs: nowMilliseconds(),
                lastActiveMs: nowMilliseconds(),
                workingDirectory: "/Users/mac/Desktop/AnyWhereVibe"
            ),
            SessionModel(
                id: "session-2",
                name: "Planning",
                status: .idle,
                createdAtMs: nowMilliseconds(),
                lastActiveMs: nowMilliseconds(),
                workingDirectory: "/Users/mac/Desktop"
            ),
        ]
    }

    private static func nowMilliseconds() -> UInt64 {
        UInt64(Date().timeIntervalSince1970 * 1_000)
    }

    private func applyAuthoritativeSessions(_ authoritativeSessions: [SessionModel]) {
        let previousActiveSessionID = activeSessionID
        sessions = authoritativeSessions

        if let previousActiveSessionID,
           authoritativeSessions.contains(where: { $0.id == previousActiveSessionID }) {
            activeSessionID = previousActiveSessionID
        } else {
            activeSessionID = authoritativeSessions.first?.id
        }
    }
}

File: android/app/src/main/java/com/mrt/app/features/sessions/SessionViewModel.kt
package com.mrt.app.features.sessions

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.mrt.app.core.models.SessionModel
import com.mrt.app.core.network.ConnectionManaging
import com.mrt.app.core.network.ConnectionState
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.CoroutineStart
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch
import mrt.Mrt
import java.util.UUID

class SessionViewModel(
    private val connectionManager: ConnectionManaging? = null,
    private val scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate),
    initialSessions: List<SessionModel> = connectionManager?.sessions?.value ?: emptyList(),
) {
    var sessions by mutableStateOf(initialSessions)
        private set
    var activeSessionId by mutableStateOf(initialSessions.firstOrNull()?.id)
        private set

    init {
        connectionManager?.let { manager ->
            scope.launch(context = Dispatchers.Unconfined, start = CoroutineStart.UNDISPATCHED) {
                manager.sessions.collectLatest { authoritativeSessions ->
                    applyAuthoritativeSessions(authoritativeSessions)
                }
            }
        }
    }

    fun selectSession(id: String): Boolean {
        if (sessions.none { it.id == id }) {
            return false
        }

        activeSessionId = id
        connectionManager?.let { manager ->
            scope.launch(start = CoroutineStart.UNDISPATCHED) {
                try {
                    manager.switchSession(sessionId = id)
                } catch (_: Throwable) {
                }
            }
        }
        return true
    }

    fun canCreateSession(connectionState: ConnectionState? = null): Boolean {
        val manager = connectionManager ?: return true
        return (connectionState ?: manager.state.value) == ConnectionState.CONNECTED
    }

    fun createSession(named: String) {
        val trimmedName = named.trim()
        if (trimmedName.isEmpty()) {
            return
        }

        connectionManager?.let { manager ->
            if (!canCreateSession()) {
                return
            }
            scope.launch(start = CoroutineStart.UNDISPATCHED) {
                try {
                    manager.createSession(name = trimmedName, workingDirectory = "")
                } catch (_: Throwable) {
                }
            }
            return
        }

        val timestamp = System.currentTimeMillis()
        val session = SessionModel(
            id = UUID.randomUUID().toString(),
            name = trimmedName,
            status = Mrt.TaskStatus.IDLE,
            createdAtMs = timestamp,
            lastActiveMs = timestamp,
            workingDirectory = "/tmp/${trimmedName.replace(" ", "-").lowercase()}",
        )
        sessions = listOf(session) + sessions
        activeSessionId = session.id
    }

    fun closeSession(id: String) {
        connectionManager?.let { manager ->
            if (sessions.none { it.id == id }) {
                return
            }
            scope.launch(start = CoroutineStart.UNDISPATCHED) {
                try {
                    manager.closeSession(sessionId = id)
                } catch (_: Throwable) {
                }
            }
            return
        }

        sessions = sessions.filterNot { it.id == id }
        activeSessionId = when {
            activeSessionId != id -> activeSessionId
            else -> sessions.firstOrNull()?.id
        }
    }

    fun cancelTask(id: String) {
        connectionManager?.let { manager ->
            if (sessions.none { it.id == id }) {
                return
            }
            scope.launch(start = CoroutineStart.UNDISPATCHED) {
                try {
                    manager.cancelTask(sessionId = id)
                } catch (_: Throwable) {
                }
            }
            return
        }

        val updatedAt = System.currentTimeMillis()
        sessions = sessions.map { session ->
            if (session.id != id) {
                session
            } else {
                session.copy(
                    status = Mrt.TaskStatus.CANCELLED,
                    lastActiveMs = updatedAt,
                )
            }
        }
    }

    private fun applyAuthoritativeSessions(authoritativeSessions: List<SessionModel>) {
        val previousSelection = activeSessionId
        sessions = authoritativeSessions
        activeSessionId = when {
            previousSelection != null && authoritativeSessions.any { it.id == previousSelection } -> previousSelection
            else -> authoritativeSessions.firstOrNull()?.id
        }
    }
}
